### ISSUE TITLE
Multilanguage support

### ISSUE THREAD

**Issue by kecinzer** [yakyak/yakyak#332]
Is there any option to add multilanguage support to YakYak?
As I saw in source, all string are hardcoded.

**Comment by davibe** [comment]
any javascript node.js i18n solution should be fine if you just want to replace strings.. bug is it really needed for our simple ui ?

**Comment by kecinzer** [comment]
If it's possible, and no power consuming, I vote for multilanguage support.

**Comment by averissimo** [comment]
Initial support in fork `i18n-multilanguae`

**Volunteers can start with translations**

Just copy the english template -- 'src/locales/en.json' -- and just add a new file to that folder.

ex: 'fr.json' or 'cs.json' :wink: :wink: @Fredouye @kecinzer  

*Template:* https://github.com/yakyak/yakyak/blob/i18n-multilanguage/src/locales/en.json

**TODO:** non-blocking though

- [ ] Take better advantage of i18n to use nesting and avoid redundant entries as plurals

**Comment by averissimo** [comment]
Preview with portuguese

![image](https://cloud.githubusercontent.com/assets/211358/21075415/95eb9204-bf09-11e6-98eb-fbc5842a978e.png)

**Comment by Fredouye** [comment]
Hi André

here's a french translation :)

http://pastebin.com/sQu5ThVF

**Comment by averissimo** [comment]
Woa!! That was fast

Two additional things:

1. Can you add the additonal translations below? (for the about-window)
1. Do you want to test this branch -- to see if any string is missing and fine tune some of the translation?
    - Can you build it yourself or do you need a binary?


```
    "A newer version is available, please upgrade from ": "A newer version is available, please upgrade from ",
    "to": "to",
    "Desktop client for Google Hangouts": "Desktop client for Google Hangouts",
    "License": "License",
    "Main authors": "Main authors",
    "Contributors": "Contributors"
```

**Comment by Fredouye** [comment]
Hi André

updated French translation : http://pastebin.com/BVqq3qUg

**Comment by averissimo** [comment]
@Fredouye I'm uploading new binaries in the 1.4.1 pre-release that have the Language support, they should be all there in about 15minutes

They have a different name `yakyak-1.4.1-i18n-experimental-...`

Please try it and see if there is any missing translation or weird behaviour :)

**Comment by Fredouye** [comment]
I'm testing this pre-release 1.4.1 on macOS 12.12.1 right now :)

One thing I've noticed : in the Edit menu, cut/copy/paste are translated in French, but not if I right click in a conversation (they're still in English).

And a small typo, line 85 : 

--- fr.json.txt.bak	2016-12-11 18:07:51.000000000 +0100
+++ fr.json.txt	2016-12-11 18:08:42.000000000 +0100
@@ -82,7 +82,7 @@
             "redo": "Rétablir",
             "paste": "Coller",
             "cut": "Couper",
-            "copy": "Coper",
+            "copy": "Copier",
             "select_all": "Tout sélectionner",
             "language": "Langage (Language)"
         },

**Comment by Fredouye** [comment]
Another small issue, reproductible in French and English : View -> Pop-Up Notification -> menu.view.notification.custom is displayed (instead of "Use YakYak custom sound for notifications"", I guess).

**Comment by averissimo** [comment]
updated the osx release. it should be fixed

**Comment by Fredouye** [comment]
Yes, you even corrected my typo :)

So far, no issue encountered.

**Comment by averissimo** [comment]
@Fredouye @davibe and anyone else that wants to weight in on this :)

what do you think would be better for the english default?

1. directly in the app (ex: `i18n.__('path.to.foo:Default string here in english for Foo')`)
1. via locale file and keep only paths in app (ex: `i18n.__('path.to.foo')`)
    - contents of en.json:

        ```
        ...
        path:
          to:
            foo: Default string here in english for Foo
        ...
        ```

**Comment by averissimo** [comment]
I almost have all the application converted to the __1.__ as I've discussed with @davibe this could be the better option

Only concern I have is with repeated statements, but there are not many.

And with the module we are using an en.yml is generated dynamically by the application

**Comment by averissimo** [comment]
pull request submited #526

**Comment by hsabo** [comment]
Hello André,
many thanks for your work.
Here is a german translation (if not too late):
[http://pastebin.com/Y3tnDe07](url)
Happy New Year,
Holger

**Comment by averissimo** [comment]
Great work Holger!!

Already added it to the i18n pull request

Happy new year!

**Comment by hsabo** [comment]
Hello André,

if I get it right, the **hl=en** in the following login URL (found in `login.coffee`) forces the Google login dialog to english language:
`LOGIN_URL = "https://accounts.google.com/o/oauth2/programmatic_auth?hl=en&scope=https%....`

What do you thing about removing the part (so using the users default language) or adding i18n feature for login also (I did not try out either one by myself)?

Besides... Thanks for the quick addition of german language!
Holger

**Comment by averissimo** [comment]
@hsabo if you can figure out how to access to `localStorage` variable (in `window`) from `login.coffee` then it should be easy. It also brings a point if switching languages should be possible here (which would require a page reload)

The other approach you talk about that uses the systems' default might create false expectations that the app is translated to a language that's not supported by YakYak.. I'm not sure if this should be the way to go but that can be easily done by calling `require('electron').app.getLocale()` and injecting on the url

Personally I think keeping an english login would be a safe bet, but if you or someone else wants to implement this I'd welcome it gladly :)

**Comment by aPoCoMiLogin** [comment]
Here is my polish translation: https://github.com/yakyak/yakyak/pull/537

**Comment by kecinzer** [comment]
Thank you for commit. I will try it ASAP.
BTW. what about plurals format? Czech language has 3 plurals - 1, 2-4, 5 and more including 0. :).

**Comment by averissimo** [comment]
great work @aPoCoMiLogin!! I've left two comments in the pull request

__Are you guys also testing it? I've been running for a while now and found few problems__

I think this should be quite ready to merge in master

**Comment by averissimo** [comment]
@kecinzer right now there is only one plural being used (notification and notifications) (and that is for interface and not dependent on number)

Have you found somewhere in the UI that required that kind of specificity in plurals?

**Comment by kecinzer** [comment]
@averissimo Ah, ok, I don't have much time to investigate, so may be it's useless :).

**Comment by averissimo** [comment]
not useless, it's just that it is not used currently in the interface.

if in the future we have a string such as `You have 5 unread messages`, then that will be required.

and the framework that we use supports plurals in the Czech language :)

**Comment by aPoCoMiLogin** [comment]
> Are you guys also testing it? I've been running for a while now and found few problems

What kind of problems?

**Comment by averissimo** [comment]
Missing translations, bad positioning, ...

**Comment by aPoCoMiLogin** [comment]
I'm sorry for not testing it on my side, cause I can't build it on osx (some issues with xcode).. I'll try to fix building it on my laptop asap

**Comment by averissimo** [comment]
don't worry, try this osx build: https://dl.dropboxusercontent.com/u/9592943/delete/yakyak-1.4.1-osx.zip

**Comment by aPoCoMiLogin** [comment]
Since saw your PR about automatic build, is there a way to set it for not only tags, but also for PR's ?

**Comment by aPoCoMiLogin** [comment]
Did tested it and it seems that it is ok.

**Comment by averissimo** [comment]
it would be cool to better test stuff out and better involve users in this process, but I don't think it is possible

documentation:
* `on` [link](https://docs.travis-ci.com/user/deployment/#Conditional-Releases-with-on%3A)
* github-releases [link](https://docs.travis-ci.com/user/deployment/releases/)

Not sure if that's something we'd want _(too much spam in releases page :) )_.. what do you think?

**Comment by aPoCoMiLogin** [comment]
Didn't mean making release, just though that there is some way that travis could post a comment or attach something to PR, it could be very cool. But just found that: https://github.com/travis-ci/travis-ci/issues/5579 so I just have to fix my setup locally for now.

**Comment by hsabo** [comment]
During my tests so far I noticed the following (regarding i18n):
- Typo in de.json: help.about.contributors should be "Mitwirkende" (not "Mitwirkende**n**")
- connection.connecting, .loading, .not_connected are not internationalized any more (think this worked in previous version, but not sure which one)
- I18n for input.message_new is missing, proposal for de.json: `"message_new": "Nachricht schreiben",`

**Comment by averissimo** [comment]
thanks @hsabo 

- Just corrected the typo (pull it and you'll see the changes)
- Those two have been corrected in affb8f11a0df79772eba62adb2e11d49b6811a71
  - We changed how the startup screen works and that translation was missing
- Corrected in the same commit (I think it was looking for the wrong 'path'.. my bad :)

**Comment by averissimo** [comment]
@aPoCoMiLogin that thing I did for automatic generation of releases by travis was still fresh in my mind and did not interpret what you were saying correctly :)

we could also set it to upload all the travis builds to some storage service, but I guess that would add up pretty quickly for free services

__let us know if you get some kind of setup going__ _(i'll be keeping tabs on https://github.com/travis-ci/travis-ci/issues/5579 as it evolves, it sounds like a great solution)_

ps. For personal builds you can always install docker and use the `dade` branch or [averissimo/docker-yakyak](https://github.com/averissimo/docker-yakyak) although both will stop being maintained if we integrate travis release

**Comment by aPoCoMiLogin** [comment]
@averissimo sorry that I didn't clarify it better.. With docker image this is some solution, it's worth mention about that image in README, cause on osx it's much easier dealing with docker, than with npm and some packages.

**Comment by averissimo** [comment]
can you create a pull request with the changes you have in mind for the readme?

multilanguage has been merged to 1.4.2 prerelease \o/ thanks for all the translations

I'm closing the issue! (although conversation can continue here, or via new issues)
